### PR TITLE
Todoist: New Source and Actions

### PR DESCRIPTION
**Pull request by michelle0927**
Addresses issues https://github.com/PipedreamHQ/pipedream/issues/1703 & https://github.com/PipedreamHQ/pipedream/issues/1495

### CODE REVIEW

**Comment by dylburger** [comment]
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Comment by vellames-turing** [comment]
Removing myself as a reviewer since I have some PRs to take a look at before it.
@js07 @jcortes Feel free to get it :)

File: components/todoist/actions/update-comment/update-comment.mjs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      propDefinition: [
+        todoist,
+        "commentId",
+      ],
+    },
+    content: {
+      propDefinition: [
+        todoist,
+        "content",
+      ],
```
Comment (js07) [inline components/todoist/actions/update-comment/update-comment.mjs:21]: Should the `content` prop be `optional: false` in this action?

**Review by js07 — COMMENTED** [review]
Looks great overall! The actions and source worked for me (aside from a prop/typo in [`uncomplete-task.mjs`](https://github.com/PipedreamHQ/pipedream/blob/b20bd8ff0457415414f1245fe9ce0e83a64ff7fe/components/todoist/actions/uncomplete-task/uncomplete-task.mjs#L14). I just made some comments on minor issues.

File: components/todoist/actions/import-tasks/import-tasks.mjs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    const fileContents = (await fs.promises.readFile(path)).toString();
+    const tasks = await converter.csv2jsonAsync(fileContents);
+    const data = tasks.map((task) => ({
+      content: task.content,
+      description: task.description,
+      project_id: project,
+      due: task.due,
+      priority: task.priority,
+      parent_id: task.parent_id,
```
Comment (js07) [inline components/todoist/actions/import-tasks/import-tasks.mjs:41]: Do we want `assignee` and `assigner` to be included in imported tasks?

```suggestion
      parent_id: task.parent_id,
      responsible_uid: task.assignee,
      assigned_by_uid: task.assigner,
```

**Review by js07 — COMMENTED** [review]
Looks good overall. Thanks for making all of those changes.

File: components/todoist/actions/import-tasks/import-tasks.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michelle0927) [reply components/todoist/actions/import-tasks/import-tasks.mjs]: Thanks. I had taken out `section_id` for the same reason, but forgot about subtasks with `parent_id`. I implemented your suggestions, and it works great.

**Review by js07 — COMMENTED** [review]
Looks good! I just made a few **minor** comments.

File: components/todoist/sources/common-task.mjs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    async getSyncResult() {
+      return this.todoist.syncItems(this.db);
+    },
+    filterResults(syncResult) {
+      return Object.values(syncResult)
+        .filter(Array.isArray)
+        .flat()
+        .filter(this.isElementRelevant)
+        .filter((element) =>
+          this.todoist.isProjectInList(element.project_id, this.selectProjects));
```
Comment (js07) [inline components/todoist/sources/common-task.mjs:28]: When I deploy or run the [`new-or-modified-task`](https://github.com/PipedreamHQ/pipedream/blob/daa0a3424a0a1352fa21704b6c4aefdf043eff2e/components/todoist/sources/new-or-modified-task/new-or-modified-task.mjs) source, and leave the `selectProjects` prop blank, I get the error
```
TypeError: Cannot read property 'length' of undefined
```
todoist-typeerror2

at this line:
https://github.com/PipedreamHQ/pipedream/blob/daa0a3424a0a1352fa21704b6c4aefdf043eff2e/components/todoist/todoist.app.mjs#L330

To fix, I'd suggest:
```suggestion
          this.todoist.isProjectInList(element.project_id, this.selectProjects ?? []));
```

File: components/todoist/todoist.app.mjs [hunk 4]
```diff
@@ -0,0 +1,1105 @@
...
+    createIfNotFound: {
+      type: "boolean",
+      label: "Create If Not Found",
+      description: "Create this item if it is not found",
+      default: false,
+    },
+    path: {
+      type: "string",
+      label: "File Path",
+      description: "Path to .csv file containing task data",
```
Comment (js07) [inline components/todoist/todoist.app.mjs:250]: Minor suggestion, to clarify how to use this prop (used in `import-tasks.mjs`):

```suggestion
      description: "Path to .csv file containing task data. Enter a static value (e.g., `/tmp/myFile.csv`) or reference prior step exports via the `steps` object (e.g., `{{steps.export_tasks.$return_value}}`).",```

**Review by js07 — COMMENTED** [review]
Looks good to me! I just found a minor bug when testing the old `new-or-modified-task` source, as mentioned a comment below.

**Review by js07 — APPROVED** [review]
Looks good. I tested the new source and actions and they all work for me.

**Comment by dannyroosevelt** [comment]
@michelle0927 plz merge when ready!

**Pull request merged**